### PR TITLE
Support custom ado fields that mark work items as duplicate

### PR DESCRIPTION
**Pull request by kananb**
## Summary of the Pull Request

Adds a new field to the ado config to specify work item fields that indicate a work item is marked as a duplicate.

## PR Checklist
* [x] Applies to work item: #3384
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [x] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

Set a bug to "Completed" with the reason "Testing", and verified that it was reopened when a bug with the same title was filed again with an ado config that included `"Sytem.Reason": "Testing"` in the "ado_duplicate_fields" property.

### CODE REVIEW

File: src/pytypes/onefuzztypes/models.py [hunk 1]
```diff
@@ -273,6 +273,7 @@ class ADOTemplate(BaseModel):
     unique_fields: List[str]
     comment: Optional[str]
     ado_fields: Dict[str, str]
+    ado_duplicate_fields: Optional[Dict[str, str]]
```
Comment (chkeita) [inline src/pytypes/onefuzztypes/models.py:276]: can you updates the examples to includes the new field
docs\notifications\ado.md
contrib\onefuzz-job-azure-devops-pipeline\ado-work-items.json

**Review by chkeita — APPROVED** [review]

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs [hunk 2]
```diff
@@ -239,7 +239,7 @@ public class Ado : NotificationsBase, IAdo {
 
         var renderedConfig = RenderAdoTemplate(logTracer, renderer, config, instanceUrl);
         var ado = new AdoConnector(renderedConfig, project!, client, instanceUrl, logTracer, await GetValidFields(client, project));
-        await ado.Process(notificationInfo);
+        await ado.Process(notificationInfo, config.AdoDuplicateFields);
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:242]: The `AdoConnector` already has an instance of the rendered config so we actually don't even need to modify the `Process` function signature.

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs [hunk 3]
```diff
@@ -535,7 +536,7 @@ public sealed class AdoConnector {
                     _logTracer.AddTags(new List<(string, string)> { ("MatchingWorkItemIds", $"{workItem.Id}") });
                     _logTracer.LogInformation("Found matching work item");
                 }
-                if (IsADODuplicateWorkItem(workItem)) {
+                if (IsADODuplicateWorkItem(workItem, duplicateFields)) {
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:539]: Continuing from earlier comment, we can pass `_config.AdoDuplicateFields` here and remove the `duplicateFields` argument.

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs [hunk 4]
```diff
@@ -575,13 +576,17 @@ public sealed class AdoConnector {
...
             // A work item could have System.State == Resolve && System.Reason == Duplicate
             // OR it could have System.State == Closed && System.Reason == Duplicate
             // I haven't found any other combinations where System.Reason could be duplicate but just to be safe
             // we're explicitly _not_ checking the state of the work item to determine if it's duplicate
             return wi.Fields.ContainsKey("System.Reason") && string.Equals(wi.Fields["System.Reason"].ToString(), "Duplicate", StringComparison.OrdinalIgnoreCase)
             || wi.Fields.ContainsKey("Microsoft.VSTS.Common.ResolvedReason") && string.Equals(wi.Fields["Microsoft.VSTS.Common.ResolvedReason"].ToString(), "Duplicate", StringComparison.OrdinalIgnoreCase)
+            || duplicateFields?.Any(fieldPair => {
+                var (field, value) = fieldPair;
+                return wi.Fields.ContainsKey(field) && string.Equals(wi.Fields[field].ToString(), value, StringComparison.OrdinalIgnoreCase);
+            }) == true
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:589]: `Any` already returns bool

Comment (kananb) [reply src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:589]: Have to add the `== true` to appease the null conditional operator here.

File: src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs (hunk 3 above)
Comment (kananb) [reply src/ApiService/ApiService/onefuzzlib/notifications/Ado.cs:539]: whipped up a quick PR with these changes
https://github.com/microsoft/onefuzz/pull/3473

**Pull request merged**